File: simulateur_v5/simulateur/game_system.py
import random
from typing import Deque, List, Optional, Tuple

from .grammar import Occurrence
from .models import Card, Pools

class SystemBrokenError(RuntimeError):
    """Deckout lose condition. Used during the refresh procedure as an emergency system."""

# All the functions used to recreate a damage sequence are written here.
class GameSystem:
    __slots__ = (
        "main_deck", "discard", "trigger_deck", "trigger_discard",
        "trigger_deck_stock", "main_deck_stock",
        "clock_zone", "level_zone", "total_damage", "initial_main_count",
        "initial_trigger_count", "initial_trigger_stock_count",
        "initial_main_stock_count",
    )

    # All card zones and check counts are initialized here
    def __init__(self, main_deck: Deque[Card], trigger_deck: Deque[Card],
                 initial_trigger_stock: Optional[List[Card]] = None,
                 initial_main_stock: Optional[List[Card]] = None):
        self.main_deck = main_deck
        self.discard: List[Card] = []
        self.trigger_deck = trigger_deck
        self.trigger_discard: List[Card] = []
        self.trigger_deck_stock: List[Card] = list(initial_trigger_stock) if initial_trigger_stock else []
        self.main_deck_stock: List[Card] = list(initial_main_stock) if initial_main_stock else []
        self.clock_zone: List[Card] = []
        self.level_zone: List[Card] = []
        self.total_damage = 0
        self.initial_main_count = len(main_deck)
        self.initial_trigger_count = len(trigger_deck)
        self.initial_trigger_stock_count = len(self.trigger_deck_stock)
        self.initial_main_stock_count = len(self.main_deck_stock)

    # Trigger deck structure

    # Safety function to verify the main deck always has a card in it
    # If both main deck and its discard are simultaneous empty, the simulator is stopped
    def _ensure_trigger_deck(self) -> None:
        if not self.trigger_deck:
            if not self.trigger_discard:
                raise SystemBrokenError("You reached the simulation's limit here")
            random.shuffle(self.trigger_discard)
            self.trigger_deck.extend(self.trigger_discard)
            self.trigger_discard.clear()

    # Takes the main deck's top card, uses it for trigger check and puts it into trigger's stock zone
    def _reveal_trigger_card(self) -> Card:
        self._ensure_trigger_deck()
        card = self.trigger_deck.pop()
        self.trigger_deck_stock.append(card)
        return card

    # Looks at the top deck, but lets it on top for later trigger checks
    def peek_trigger_top(self) -> Card:
        self._ensure_trigger_deck()
        return self.trigger_deck[-1]

    @staticmethod
    # Failsafe function here, actually there is no card in the game that has more than two soul triggers in it
    def __card_soul(card: Card) -> int:
        return card.n_triggers if card.n_triggers <= 2 else 0

    # Trigger check rule
    def trigger_check(self) -> int:
        card = self._reveal_trigger_card()
        return self.__card_soul(card)

    # Double trigger check rule
    def twin_drive_check(self) -> int:
        total = 0
        for _ in range(2):
            card = self._reveal_trigger_card()
            total += self.__card_soul(card)
        return total

    # Damage structure

    # Level up rule
    # While the clock zone holds 7 or more cards, process a batch of 7:
    # move the first non-climax card found in that batch to the level zone, and discard the rest of the batch.
    # in case there are 7 climaxes in clock zone, one is forced to go into level zone
    def levelup(self) -> None:
        while len(self.clock_zone) >= 7:
            batch = [self.clock_zone.pop(0) for _ in range(7)]
            level_card_idx = next(
            (i for i, card in enumerate(batch) if card.category != "climax"),
            None,)
            if level_card_idx is not None:
                self.level_zone.append(batch.pop(level_card_idx))
            self.discard.extend(batch)

    # Refresh penalty rule
    def refresh(self) -> None:
        if not self.discard:
            raise SystemBrokenError("You reached the simulation's limit here")
        new_deck = self.discard[:]
        self.discard.clear()
        random.shuffle(new_deck)
        self.clock_zone.append(new_deck.pop())
        self.total_damage += 1
        self.levelup()
        self.main_deck.extend(new_deck)

    def keep_cards(self, cards: List[Card]) -> None:
        self.clock_zone.extend(cards)
        self.total_damage += len(cards)
        self.levelup()

    # Cost primitive

    def can_pay_trigger(self, n: int) -> bool:
        return len(self.trigger_deck_stock) >= n

    def pay_trigger(self, n: int) -> None:
        for _ in range(n):
            self.trigger_discard.append(self.trigger_deck_stock.pop())

    def can_pay_main(self, n: int) -> bool:
        return len(self.main_deck_stock) >= n

    def pay_main(self, n: int) -> None:
        for _ in range(n):
            self.discard.append(self.main_deck_stock.pop())

    # Burn primitive

    def reveal_trigger_cards(self, n: int) -> Tuple[int, List[Card]]:
        cards = [self._reveal_trigger_card() for _ in range(n)]
        bonus = sum(self.__card_soul(c) for c in cards)
        return bonus, cards

    def resolution_core(self, n: int) -> bool:
        pile: List[Card] = []
        canceled = False
        while len(pile) < n:
            if not self.main_deck:
                self.refresh()
                continue
            card = self.main_deck.pop()
            pile.append(card)
            if card.category == "climax":
                self.discard.extend(pile)
                canceled = True
                break

        if not canceled:
            self.keep_cards(pile)

        return canceled

    def resolution(self, n: int, trigger: Optional[str]) -> bool:
        bonus = 0
        if trigger == "single":
            bonus, _ = self.reveal_trigger_cards(1)
        elif trigger == "twin":
            bonus, _ = self.reveal_trigger_cards(2)
        return self.resolution_core(n + bonus)

    # Mill primitive

    def mill(self, n: int, target: str, edge: str) -> Tuple[int, List[Card]]:
        if target == "self":
            deck, discard_pile = self.trigger_deck, self.trigger_discard
        else:
            deck, discard_pile = self.main_deck, self.discard

        milled: List[Card] = []
        climax_count = 0
        while len(milled) < n:
            if not deck:
                if target == "self":
                    self._ensure_trigger_deck()
                else:
                    self.refresh()
                continue
            card = deck.pop() if edge == "top" else deck.popleft()
            discard_pile.append(card)
            milled.append(card)
            if card.category == "climax":
                climax_count += 1
        return climax_count, milled

    # Scry primitive

    def scry(self, n: int) -> Tuple[int, List[Card]]:
        count = min(n, len(self.main_deck))
        revealed = [self.main_deck.pop() for _ in range(count)]

        climax_cards = [c for c in revealed if c.category == "climax"]
        kept = [c for c in revealed if c.category != "climax"]
        self.discard.extend(climax_cards)
        for card in reversed(kept):
            self.main_deck.append(card)

        return len(climax_cards), revealed

    # Shuffle primitive

    def shuffle_into_main(self, moved: List[Card], kept: List[Card]) -> None:
        self.discard = kept
        new_main = list(self.main_deck) + moved
        random.shuffle(new_main)
        self.main_deck.clear()
        self.main_deck.extend(new_main)

    def shuffle_discard_into_main(self, count: int) -> None:
        eligible_idx = [i for i, c in enumerate(self.discard) if c.category != "climax"]
        chosen_idx = set(random.sample(eligible_idx, min(count, len(eligible_idx))))
        if not chosen_idx:
            return
        moved = [c for i, c in enumerate(self.discard) if i in chosen_idx]
        kept = [c for i, c in enumerate(self.discard) if i not in chosen_idx]
        self.shuffle_into_main(moved, kept)

    # StockSwap primitive

    def recomplete(self, n: int) -> List[Card]:
        drawn: List[Card] = []
        while len(drawn) < n:
            if not self.main_deck:
                self.refresh()
                continue
            drawn.append(self.main_deck.pop())
        return drawn

    def stock_swap(self) -> None:
        n = len(self.main_deck_stock)
        self.discard.extend(self.main_deck_stock)
        self.main_deck_stock.clear()
        self.main_deck_stock.extend(self.recomplete(n))

    def stock_shuffle(self) -> None:
        n = len(self.main_deck_stock)

        new_main = list(self.main_deck) + self.main_deck_stock
        self.main_deck_stock.clear()
        random.shuffle(new_main)
        self.main_deck.clear()
        self.main_deck.extend(new_main)

        self.main_deck_stock.extend(self.recomplete(n))

    def shuffle_all_but_climax_reserve(self, x: int) -> None:
        climax_idx = [i for i, c in enumerate(self.discard) if c.category == "climax"]
        hold_back_count = min(x, len(climax_idx))
        held_idx = set(random.sample(climax_idx, hold_back_count))

        moved = [c for i, c in enumerate(self.discard) if i not in held_idx]
        kept = [c for i, c in enumerate(self.discard) if i in held_idx]
        self.shuffle_into_main(moved, kept)

    
    def run(self, occurrences: List[Occurrence]) -> int:
        for occurrence in occurrences:
            occurrence.resolve(self)
        return self.total_damage

    # Cards tracked, used to look for bugs and data loss
    def check_invariant(self) -> None:
        tracked = (len(self.main_deck) + len(self.discard)
                   + len(self.clock_zone) + len(self.level_zone)
                   + len(self.main_deck_stock))
        expected = self.initial_main_count + self.initial_main_stock_count
        if tracked != expected:
            raise AssertionError(
                f"Card conservation inconsistency: {tracked} cards "
                f"tracked (main_deck+discard+clock_zone+level_zone"
                f"+main_deck_stock) for {expected} at the start "
                f"(initial_main_count+initial_main_stock_count)."
            )

        trigger_tracked = (len(self.trigger_deck) + len(self.trigger_discard)
                            + len(self.trigger_deck_stock))
        trigger_expected = self.initial_trigger_count + self.initial_trigger_stock_count
        if trigger_tracked != trigger_expected:
            raise AssertionError(
                f"Trigger card conservation inconsistency: "
                f"{trigger_tracked} cards tracked "
                f"(trigger_deck+trigger_discard+trigger_deck_stock) "
                f"for {trigger_expected} at the start "
                f"(initial_trigger_count+initial_trigger_stock_count)."
            )

def run_trial_from_decks(main_deck: Deque[Card], trigger_deck: Deque[Card],
                          trigger_deck_stock: List[Card], main_deck_stock: List[Card],
                          occurrences: List[Occurrence], check_invariants: bool = False) -> int:
    state = GameSystem(main_deck, trigger_deck, trigger_deck_stock, main_deck_stock)
    damage = state.run(occurrences)
    if check_invariants:
        state.check_invariant()
    return damage


def run_single_trial(deck_size, climax_n, occurrences: List[Occurrence], pools: Pools,
                      trigger_deck_size, trigger_deck_climax,
                      trigger_deck_stock_size: int = 0,
                      main_deck_stock_size: int = 0,
                      check_invariants=False) -> int:
    from .deck import build_main_deck_and_stock, build_trigger_deck_and_stock

    main_deck, main_deck_stock = build_main_deck_and_stock(
        pools.main_climax, pools.main_non_climax, deck_size, climax_n,
        main_deck_stock_size,
    )
    trigger_deck, trigger_deck_stock = build_trigger_deck_and_stock(
        pools.trigger_climax, pools.trigger_non_climax,
        trigger_deck_size, trigger_deck_climax, trigger_deck_stock_size,
    )
    return run_trial_from_decks(
        main_deck, trigger_deck, trigger_deck_stock, main_deck_stock,
        occurrences, check_invariants,
    )

File: js_version/js/gameSystem.js
(function () {
// gameSystem.js — port of game_system.py
// main_deck / trigger_deck are plain JS arrays used as stacks: push()/pop()
// operate on the "top" of the deck (matches Python's deque.append/pop on the
// right end). Mill's "bottom" edge uses shift() (matches deque.popleft()).

function shuffleInPlace(arr, rng) {
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr;
}

/** Sample k unique indices from [0, n) without replacement. */
function sampleIndices(n, k, rng) {
  const idx = Array.from({ length: n }, (_, i) => i);
  shuffleInPlace(idx, rng);
  return idx.slice(0, k);
}

class SystemBrokenError extends Error {}

class GameSystem {
  constructor(mainDeck, triggerDeck, initialTriggerStock = [], initialMainStock = [], rng = Math.random) {
    this.rng = rng;
    this.mainDeck = mainDeck.slice();
    this.discard = [];
    this.triggerDeck = triggerDeck.slice();
    this.triggerDiscard = [];
    this.triggerDeckStock = initialTriggerStock.slice();
    this.mainDeckStock = initialMainStock.slice();
    this.clockZone = [];
    this.levelZone = [];
    this.totalDamage = 0;
    this.initialMainCount = this.mainDeck.length;
    this.initialTriggerCount = this.triggerDeck.length;
    this.initialTriggerStockCount = this.triggerDeckStock.length;
    this.initialMainStockCount = this.mainDeckStock.length;
  }

  shuffle(arr) { return shuffleInPlace(arr, this.rng); }
  sample(n, k) { return sampleIndices(n, k, this.rng); }

  // -- Trigger deck structure --

  ensureTriggerDeck() {
    if (this.triggerDeck.length === 0) {
      if (this.triggerDiscard.length === 0) throw new SystemBrokenError("congrats, you just broke the system");
      this.shuffle(this.triggerDiscard);
      this.triggerDeck.push(...this.triggerDiscard);
      this.triggerDiscard.length = 0;
    }
  }

  revealTriggerCard() {
    this.ensureTriggerDeck();
    const card = this.triggerDeck.pop();
    this.triggerDeckStock.push(card);
    return card;
  }

  peekTriggerTop() {
    this.ensureTriggerDeck();
    return this.triggerDeck[this.triggerDeck.length - 1];
  }

  static cardSoul(card) { return card.nTriggers <= 2 ? card.nTriggers : 0; }

  triggerCheck() {
    const card = this.revealTriggerCard();
    return GameSystem.cardSoul(card);
  }

  twinDriveCheck() {
    let total = 0;
    for (let i = 0; i < 2; i++) total += GameSystem.cardSoul(this.revealTriggerCard());
    return total;
  }

  // -- Damage structure --

  levelup() {
    while (this.clockZone.length >= 7) {
      const batch = this.clockZone.splice(0, 7);
      const idx = batch.findIndex((c) => c.category !== "climax");
      if (idx !== -1) this.levelZone.push(batch.splice(idx, 1)[0]);
      this.discard.push(...batch);
    }
  }

  refresh() {
    if (this.discard.length === 0) throw new SystemBrokenError("congrats, you just reached the deckout lose condition");
    const newDeck = this.discard.slice();
    this.discard.length = 0;
    this.shuffle(newDeck);
    this.clockZone.push(newDeck.pop());
    this.totalDamage += 1;
    this.levelup();
    this.mainDeck.push(...newDeck);
  }

  keepCards(cards) {
    this.clockZone.push(...cards);
    this.totalDamage += cards.length;
    this.levelup();
  }

  // -- Cost primitives --

  canPayTrigger(n) { return this.triggerDeckStock.length >= n; }
  payTrigger(n) { for (let i = 0; i < n; i++) this.triggerDiscard.push(this.triggerDeckStock.pop()); }
  canPayMain(n) { return this.mainDeckStock.length >= n; }
  payMain(n) { for (let i = 0; i < n; i++) this.discard.push(this.mainDeckStock.pop()); }

  // -- Burn primitive --

  revealTriggerCards(n) {
    const cards = [];
    for (let i = 0; i < n; i++) cards.push(this.revealTriggerCard());
    const bonus = cards.reduce((s, c) => s + GameSystem.cardSoul(c), 0);
    return { bonus, cards };
  }

  resolutionCore(n) {
    const pile = [];
    let canceled = false;
    while (pile.length < n) {
      if (this.mainDeck.length === 0) { this.refresh(); continue; }
      const card = this.mainDeck.pop();
      pile.push(card);
      if (card.category === "climax") {
        this.discard.push(...pile);
        canceled = true;
        break;
      }
    }
    if (!canceled) this.keepCards(pile);
    return canceled;
  }

  resolution(n, trigger) {
    let bonus = 0;
    if (trigger === "single") ({ bonus } = this.revealTriggerCards(1));
    else if (trigger === "twin") ({ bonus } = this.revealTriggerCards(2));
    return this.resolutionCore(n + bonus);
  }

  // -- Mill primitive --

  mill(n, target, edge) {
    const isSelf = target === "self";
    const deck = isSelf ? this.triggerDeck : this.mainDeck;
    const discardPile = isSelf ? this.triggerDiscard : this.discard;

    const milled = [];
    let climaxCount = 0;
    while (milled.length < n) {
      if (deck.length === 0) {
        if (isSelf) this.ensureTriggerDeck();
        else this.refresh();
        continue;
      }
      const card = edge === "top" ? deck.pop() : deck.shift();
      discardPile.push(card);
      milled.push(card);
      if (card.category === "climax") climaxCount++;
    }
    return { climaxCount, milled };
  }

  // -- Scry primitive --

  scry(n) {
    const count = Math.min(n, this.mainDeck.length);
    const revealed = [];
    for (let i = 0; i < count; i++) revealed.push(this.mainDeck.pop());

    const climaxCards = revealed.filter((c) => c.category === "climax");
    const kept = revealed.filter((c) => c.category !== "climax");
    this.discard.push(...climaxCards);
    for (let i = kept.length - 1; i >= 0; i--) this.mainDeck.push(kept[i]);

    return { climaxCount: climaxCards.length, revealed };
  }

  // -- Shuffle primitive --

  shuffleIntoMain(moved, kept) {
    this.discard = kept;
    const newMain = this.mainDeck.concat(moved);
    this.shuffle(newMain);
    this.mainDeck.length = 0;
    this.mainDeck.push(...newMain);
  }

  shuffleDiscardIntoMain(count) {
    const eligibleIdx = [];
    this.discard.forEach((c, i) => { if (c.category !== "climax") eligibleIdx.push(i); });
    const chosen = new Set(this.sample(eligibleIdx.length, Math.min(count, eligibleIdx.length)).map((i) => eligibleIdx[i]));
    if (chosen.size === 0) return;
    const moved = this.discard.filter((c, i) => chosen.has(i));
    const kept = this.discard.filter((c, i) => !chosen.has(i));
    this.shuffleIntoMain(moved, kept);
  }

  // -- StockSwap primitive --

  recomplete(n) {
    const drawn = [];
    while (drawn.length < n) {
      if (this.mainDeck.length === 0) { this.refresh(); continue; }
      drawn.push(this.mainDeck.pop());
    }
    return drawn;
  }

  stockSwap() {
    const n = this.mainDeckStock.length;
    this.discard.push(...this.mainDeckStock);
    this.mainDeckStock.length = 0;
    this.mainDeckStock.push(...this.recomplete(n));
  }

  stockShuffle() {
    const n = this.mainDeckStock.length;
    const newMain = this.mainDeck.concat(this.mainDeckStock);
    this.mainDeckStock.length = 0;
    this.shuffle(newMain);
    this.mainDeck.length = 0;
    this.mainDeck.push(...newMain);
    this.mainDeckStock.push(...this.recomplete(n));
  }

  shuffleAllButClimaxReserve(x) {
    const climaxIdx = [];
    this.discard.forEach((c, i) => { if (c.category === "climax") climaxIdx.push(i); });
    const holdBackCount = Math.min(x, climaxIdx.length);
    const held = new Set(this.sample(climaxIdx.length, holdBackCount).map((i) => climaxIdx[i]));
    const moved = this.discard.filter((c, i) => !held.has(i));
    const kept = this.discard.filter((c, i) => held.has(i));
    this.shuffleIntoMain(moved, kept);
  }

  run(occurrences) {
    for (const occurrence of occurrences) occurrence.resolve(this);
    return this.totalDamage;
  }

  checkInvariant() {
    const tracked = this.mainDeck.length + this.discard.length + this.clockZone.length
      + this.levelZone.length + this.mainDeckStock.length;
    const expected = this.initialMainCount + this.initialMainStockCount;
    if (tracked !== expected) {
      throw new Error(
        `Card conservation inconsistency: ${tracked} cards tracked (mainDeck+discard+clockZone+levelZone` +
        `+mainDeckStock) for ${expected} at the start (initialMainCount+initialMainStockCount).`
      );
    }

    const triggerTracked = this.triggerDeck.length + this.triggerDiscard.length + this.triggerDeckStock.length;
    const triggerExpected = this.initialTriggerCount + this.initialTriggerStockCount;
    if (triggerTracked !== triggerExpected) {
      throw new Error(
        `Trigger card conservation inconsistency: ${triggerTracked} cards tracked ` +
        `(triggerDeck+triggerDiscard+triggerDeckStock) for ${triggerExpected} at the start ` +
        `(initialTriggerCount+initialTriggerStockCount).`
      );
    }
  }
}

function runTrialFromDecks(mainDeck, triggerDeck, triggerDeckStock, mainDeckStock, occurrences, checkInvariants = false, rng = Math.random) {
  const state = new GameSystem(mainDeck, triggerDeck, triggerDeckStock, mainDeckStock, rng);
  const damage = state.run(occurrences);
  if (checkInvariants) state.checkInvariant();
  return damage;
}

const _exports = { GameSystem, SystemBrokenError, runTrialFromDecks, shuffleInPlace, sampleIndices };
if (typeof module !== "undefined") {
  module.exports = _exports;
} else {
  window.TCG = window.TCG || {};
  Object.assign(window.TCG, _exports);
}

})();
